draft-ietf-mls-protocol.md

```diff
@@ -1435,48 +1461,226 @@ all arrive at the following state:
-derive key material for symmetric operations such as encryption/AEAD and MAC;
-AEAD or MAC MUST be used to ensure that the message originated from a member
-of the group.
+~~~~~
 
-In addition, asymmetric signatures SHOULD be used to authenticate the sender
-of a message.
+The Application context provided together with the previous Application secret
+is used to bind the Application messages with the next key and add some freshness.
 
-In combination with server-side enforced ordering, data from previous messages
-is used (as a salt when hashing) to:
+[[OPEN ISSUE: Context is left empty for now, it might be that using only
+the message counter is enough, this would be more convenient.
+Hashing all the data is obviously very costly and prevents from encrypt in
+parallel, an other solution could be to add a pseudo-random contribution to
+each message and only hash these. ]]
 
- * add freshness to derived symmetric keys
- * cryptographically bind the transcript of all previous messages with the current
-   group shared secret
+[[ OPEN ISSUE: At the moment there is no contributivity of Application secrets
+chained from the initial one to the next generation of Epoch secret. While this
+seems safe because cryptographic operations using the application secrets can't
+affect the group init_secret, it remains to be proven correct. ]]
 
-Possible candidates for that are:
+### Updating the Application Secret
 
- * the key used for the previous message (hash ratcheting)
- * the counter of the previous message (needs to be known to new members of the group)
- * the hash of the previous message (proof that other participants saw the same history)
+The following rules apply to an Application Secret:
 
-The requirement for this is that all participants know these values.
-If additional clear-text fields are attached to messages (like the counter), those
-fields MUST be protected by a signed message envelope.
+- Senders MUST only use the Application Secret once and monotonically
+  increment the generation of their secret.
+
+- Receivers SHOULD delete an Application Secret once it has been used to
+  derive the corresponding AEAD key and nonce and MAY keep those around
+  for some reasonable period.
+
+- Receivers MUST delete an Application Secret once it has been used to
+  successfully decrypt a message.
+
+[[ OPEN ISSUE: A stronger requirement could be a MUST delete Application Secret N
+as soon as N+1 has been derived.]]
+
+### Application AEAD Key Calculation
+
+The Application AEAD keying material is generated from the following
+input values:
+
+- The Application Secret value;
+- A purpose value indicating the specific value being generated;
+- The length of the key being generated.
+
+The traffic keying material is generated from an input traffic secret value using:
+
+~~~~
+write_nonce_[sender] =
+  HKDF-Expand-Label(Application_Secret,"nonce", "", nonce_length)
+
+write_key_[sender] =
+  HKDF-Expand-Label(Application_Secret,"key", "", key_length)
```

Comment:
This seems to duplicate the daigram above.

Reply:
Removed it.